### PR TITLE
fix(router/scheduler): replace the occurrences of gen_range with a safer alternative

### LINKED ISSUE #1754: [BUGFIX] Replacing the occurrences of `gen_range` with a safe alternative
## Description

Following up with the issue https://github.com/rust-random/rand/issues/1326 created by @lsampras. The function `gen_range` in itself can cause panics and error handling isn't done for some cases. This is a unpredictable behaviour and needs to be replaced with a safe alternative which allows for error handling and recoverability.

## Objective

- Deciding open a safe alternative for `gen_range` while keeping the behaviour relatively same
- Replacing the occurrences of `gen_range` with the decided safe alternative

## Guidelines

- Please mention the approach for solving the above mentioned issue and reasoning of it being recoverable.

### PR DESCRIPTION
**Pull request by anupj**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This is a bugfix for issue [1754](https://github.com/juspay/hyperswitch/issues/1754).
I have replaced the instances of `gen_range()` with `Uniform.sample`.
This should reduce the risk of `gen_range()` throwing a `panic!`.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

Provide links to the files with corresponding changes.

Following are the paths where you can find the changed files:
1. crates/router/src/routes/dummy_connector/utils.rs
2. crates/router/src/scheduler/producer.rs
3. crates/router/src/scheduler/consumer.rs
4. crates/scheduler/src/errors.rs

## Motivation and Context

This is a bugfix for issue [1754](https://github.com/juspay/hyperswitch/issues/1754)
Closes #1754.

## How did you test it?
I did not write any new tests but I ran `cargo test -p router --lib` and all tests passed.

NOTE: I was not able to run all integration tests because my laptop is not setup for running integration tests. 

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/scheduler/src/producer.rs [hunk 1]
```diff
@@ -25,9 +25,15 @@ pub async fn start_producer<T>(
...
-    let timeout = rand::thread_rng().gen_range(0..=scheduler_settings.loop_interval);
-    tokio::time::sleep(std::time::Duration::from_millis(timeout)).await;
+    use std::time::Duration;
+
+    use rand::distributions::{Distribution, Uniform};
+
+    let mut rng = rand::thread_rng();
+    let timeout = Uniform::try_from(0..=scheduler_settings.loop_interval)
+        .into_report()
+        .change_context(errors::ProcessTrackerError::InvalidInput)?;
```
Comment (SanchithHegde) [inline crates/scheduler/src/producer.rs:35]: Could you please raise `ConfigurationError` instead, as suggested in https://github.com/juspay/hyperswitch/pull/1908#discussion_r1296186188? (And remove the `InvalidInput` enum variant?)

File: crates/scheduler/src/consumer.rs [hunk 2]
```diff
@@ -37,10 +38,13 @@ pub async fn start_consumer<T: SchedulerAppState + 'static>(
...
 
-    use rand::Rng;
+    use rand::distributions::{Distribution, Uniform};
 
-    let timeout = rand::thread_rng().gen_range(0..=settings.loop_interval);
-    tokio::time::sleep(Duration::from_millis(timeout)).await;
+    let mut rng = rand::thread_rng();
+    let timeout = Uniform::try_from(0..=settings.loop_interval)
+        .into_report()
+        .change_context(errors::ProcessTrackerError::InvalidInput)?;
```
Comment (SanchithHegde) [inline crates/scheduler/src/consumer.rs:46]: Same here.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

(For anyone else landing on this PR, previous conversation happened on #1908.)

**Review by SanchithHegde — APPROVED** [review]
Looks good to me!

**Review by lsampras — APPROVED** [review]

**Pull request merged**